### PR TITLE
✨ Add ISCN gas limit estimation in case big size tx

### PR DESCRIPTION
**Pull request by shellyYG**

### CODE REVIEW

File: util/cosmos/iscn/sign.js [hunk 1]
```diff
@@ -9,11 +9,10 @@ import {
 } from '@cosmjs/stargate';
 import jsonStringify from 'fast-json-stable-stringify';
 
-import { DEFAULT_GAS_PRICE_NUMBER } from '../../CosmosHelper';
+import { getGasLimitParams } from '@/util/cosmos/iscn/gasEstimator';
```
Comment (williamchong) [inline util/cosmos/iscn/sign.js:12]: Should just return the result of the linear regression as hardcode constant in this repo to save computation power of client

File: util/cosmos/iscn/sign.js [hunk 2]
```diff
@@ -179,17 +178,50 @@ export async function estimateISCNTxFee(tx, {
...
+    },
+  };
+  const obj = {
+    type: 'cosmos-sdk/StdTx',
+    value: Buffer.from(jsonStringify(value), 'utf-8'),
+  };
+  const GAS_BUFFER = 50000;
+  const gasBuffer = BigNumber(GAS_BUFFER);
+  const { slop, intercept } = await getGasLimitParams();
+  const interceptWithBuffer = BigNumber(intercept).plus(gasBuffer);
```
Comment (williamchong) [inline util/cosmos/iscn/sign.js:211]: .plus support string input, so no need to cast GAS_BUFFER into BigNumber

File: util/cosmos/iscn/gasEstimator.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  '074E8B35B54112F23F53D227220096E14EFD431D7EE532526BF6C08F1E43D6D2',
+  '58387DB7BB02F71EF00F7415B27A6D07C2968AB8843CFCAB4DAB3DE95D83454A',
+  '5D469C6FFD1B6F872C3B49CBFE5CB77A7085CD6F11EE6DAF25B11638F8109BCB',
+  '9032C69A3A56D419BA430455D812602C6605F30B8A4A809E276D06145CA33139',
+  '58CB52A2EB17977C4B358E6AF6F35ED943FBFC961A24906708963431BC5847D4',
+  'A02E2C93BB935C57CA189C7351B82037EC9C45BBA1A3EB3B9D9084C8C9A1C0B7',
+  'AF77DEB505FEB9BEAC1C1FFD44CBADBBAEDF178B0DF88A020FA3A66600A8896D',
+  '5C2818D77AD6481234C70AF40A7336B26100176F16F970D036C9FC726CE0D2D2',
+  '12A7DEC5E75CB622567AC688BBBDF1D74CAA39E78233B4B9210640E5D2AE0138',
+]
```
Comment (williamchong) [inline util/cosmos/iscn/gasEstimator.js:36]: Should not include this tx simulation script in this production site repo

File: store/modules/actions/payment.js [hunk 4]
```diff
@@ -80,6 +81,7 @@ export async function calculateISCNTxTotalFee({ commit },
     url,
   } = payload;
   const { ISCNTotalFee } = await calculateISCNTotalFee({
+    from,
```
Comment (williamchong) [inline store/modules/actions/payment.js:84]: We should use `cosmosWallet` (or `from` if that is better )for all input for consistency, but not both

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/in/widget/iscn/index.vue [hunk 5]
```diff
@@ -374,6 +374,7 @@ export default {
           url,
         } = this;
         const txHash = await this.sendISCNSignature({
+          cosmosWallet,
           userId: this.getUserId,
           displayName: this.getUserInfo.displayName,
           from,
```
Comment (williamchong) [inline pages/in/widget/iscn/index.vue:380]: Is `from` or `cosmosWallet` actually used in this function?

File: store/modules/actions/payment.js [hunk 6]
```diff
@@ -128,6 +130,8 @@ export async function sendISCNSignature(
       type,
       license,
       publisher,
+      description,
+      cosmosWallet,
```
Comment (williamchong) [inline store/modules/actions/payment.js:134]: Should remove the `from` above if its actually not used in the later flow

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/in/widget/iscn/index.vue (hunk 5 above)
Comment (shellyYG) [reply pages/in/widget/iscn/index.vue:380]: it's cosmosWallet, will change

File: util/cosmos/iscn/sign.js [hunk 7]
```diff
@@ -179,17 +182,47 @@ export async function estimateISCNTxFee(tx, {
...
+  const interceptWithBuffer = BigNumber(GAS_ESTIMATOR_INTERCEPT).plus(GAS_ESTIMATOR_BUFFER);
+  const txBytes = Buffer.from(jsonStringify(obj), 'utf-8');
+  const byteSize = BigNumber(txBytes.length);
+  const gasUsedEstimation = byteSize.multipliedBy(GAS_ESTIMATOR_SLOP).plus(interceptWithBuffer);
   return {
     gasFee: {
-      amount: [{ amount: (DEFAULT_GAS_PRICE_NUMBER * ISCN_GAS).toFixed(), denom: 'nanolike' }],
-      gas: ISCN_GAS.toFixed(),
+      amount: [{ amount: parseInt(gasUsedEstimation.toFixed(0), 10), denom: 'nanolike' }],
+      gas: gasUsedEstimation.toFixed(0),
```
Comment (williamchong) [inline util/cosmos/iscn/sign.js:219]:
```suggestion
      amount: [{ amount: gasUsedEstimation.toFixed(), denom: 'nanolike' }],
      gas: gasUsedEstimation.toFixed(),
```
Should return bignumber or string for nanolike for accuracy, string here is better for consistency

**Review by williamchong — CHANGES_REQUESTED** [review]

**Review by williamchong — APPROVED** [review]

**Pull request merged**